Content.Shared/Turnstile/Systems/SharedTurnstileSystem.cs

```diff
@@ -0,0 +1,239 @@
+        if (!Tags.HasTag(args.OtherEntity, "DoorBumpOpener"))
+            return;
+
+        // Check the contact normal against our direction.
+        // For simplicity, we always want a mob to pass from the "back" to the "front" of the turnstile.
+        // That allows unanchored turnstiles to be dragged and rotated as needed, and will admit passage in the
+        // direction that they are pulled in.
+        var turnstile = ent.Comp;
+        if (turnstile.State == TurnstileState.Idle)
+        {
+            var facingDirection = GetFacingDirection(ent.Owner, turnstile);
+            var directionOfContact = GetDirectionOfContact(ent.Owner, args.OtherEntity);
+            if (facingDirection == directionOfContact)
+            {
+                // Admit the entity.
+                var comp = EnsureComp<PreventCollideComponent>(ent);
+                comp.Uid = args.OtherEntity;
+
+                // Play sound of turning
+                Audio.PlayPvs(turnstile.TurnSound, ent.Owner, AudioParams.Default.WithVolume(-3));
+
+                // We have to set collidable to false for one frame, otherwise the client does not
+                // predict the removal of the contacts properly, and instead thinks the mob is colliding when it isn't.
+                SetCollidable(ent.Owner, false);
+                SetState(ent.Owner, TurnstileState.Rotating);
+            }
+            else
+            {
+                // Reject the entity, play sound with cooldown
+                Audio.PlayPvs(turnstile.BumpSound, ent.Owner, AudioParams.Default.WithVolume(-3));
+            }
+        }
+        else
+        {
+            // Reject the entity, play sound with cooldown
+            Audio.PlayPvs(turnstile.BumpSound, ent.Owner, AudioParams.Default.WithVolume(-3));
+        }
+    }
+
+    /// <summary>
+    ///     Makes a turnstile proceed to the next state.
+    /// </summary>
+    private void NextState(Entity<TurnstileComponent> ent, TimeSpan time)
+    {
+        var turnstile = ent.Comp;
+
+        if (turnstile.State == TurnstileState.Rotating)
+        {
+            // Check and see if we're still colliding with the admitted entity.
+            var stillCollidingWithAdmitted = GetIsCollidingWithAdmitted(ent.Owner);
+            if (!stillCollidingWithAdmitted)
+            {
+                turnstile.NextStateChange = null;
+                SetState(ent.Owner, TurnstileState.Idle);
+                RemComp<PreventCollideComponent>(ent.Owner);
+            }
+            else
+            {
+                turnstile.NextStateChange = GameTiming.CurTime + turnstile.TurnstileTurnTime;
+            }
+        }
+    }
+
+    private void OnRemove(Entity<TurnstileComponent> turnstile, ref ComponentRemove args)
+    {
+        _activeTurnstiles.Remove(turnstile);
+    }
+
+    private bool GetIsCollidingWithAdmitted(EntityUid uid, TurnstileComponent? turnstile = null, PreventCollideComponent? preventCollide = null)
+    {
```

Comment:
This should take `Entity<TurnstileComponent?, PreventCollideComponent?>` instead.